File: cpp/problems/118-pascals-triangle.cpp
class Solution {
 public:
  vector<vector<int>> generate(int numRows) {
    // method 1
    // time complexity: O(n^2)
    // space complexity: O(n^2)

    vector<vector<int>> matrix(numRows);

    for (int i = 0; i < numRows; ++i) {
      matrix[i].resize(i + 1);
      matrix[i][0] = matrix[i][i] = 1;

      for (int j = 1; j < i; ++j) {
        matrix[i][j] = matrix[i - 1][j - 1] + matrix[i - 1][j];
      }
    }

    return matrix;
  }
};

File: js/problems/118-pascals-triangle.js
/**
 * @param {number} numRows
 * @return {number[][]}
 */
var generate = function (numRows) {
  const dp = [[1]];

  for (let i = 0; i < numRows - 1; i++) {
      let prevRow = dp[i];
      let cur = [1];

      for (let j = 0; j < prevRow.length - 1; j++) {
          cur[j + 1] = prevRow[j] + prevRow[j + 1];
      }

      cur.push(1);

      dp.push(cur);
  }

  return dp;
};

console.log(generate(3));